ocaml/xapi/xapi_session.ml

```diff
@@ -27,6 +27,228 @@ open Client
+     been a failed login attempt since the last time the stats
+     were queried, you won't get any stats *)
+  val get_stats_string : unit -> string option
+
+  val on_fail :
+       __context:Context.t
+    -> now:Date.iso8601
+    -> uname:string option
+    -> originator:string option
+    -> record:[> `log_only | `log_and_alert]
+    -> unit
+end = struct
+  type client = {
+      user_agent: string option
+    ; uname: string option
+    ; originator: string option
+  }
+
+  let client_of_info ~__context ~originator ~uname =
+    let user_agent = Context.get_user_agent __context in
+    (* check to make sure we have at least _some_ information *)
+    if
+      [user_agent; originator; uname]
+      |> List.for_all (function None | Some "" -> true | _ -> false)
+    then
+      None
+    else
+      Some {originator; uname; user_agent}
+
+  let string_of_client x =
+    [
+      ("User-Agent", x.user_agent)
+    ; ("Username", x.uname)
+    ; ("Originator", x.originator)
+    ]
+    |> List.filter_map (fun (label, value) ->
+           match value with
+           | None | Some "" ->
+               None
+           | Some value ->
+               Some (Printf.sprintf "%s=%s" label value))
+    |> String.concat ", "
+
+  type client_failed_attempts = {
+      client: client
+    ; num_failed_attempts: int
+    ; last_failed_attempt: Date.iso8601
+  }
+
+  type stats = {
+      total_num_failed_attempts: int
+    ; top_3_worst_clients: client_failed_attempts list (* has length <= 3 *)
+    ; unknown_client_failed_attempts: int
+  }
+
+  let string_of_stats
+      {
+        total_num_failed_attempts
+      ; top_3_worst_clients
+      ; unknown_client_failed_attempts
+      } =
+    let string_of_client_failed_attempts x =
+      Printf.sprintf "- %i failed login attempts! Last failed attempt: %s, %s"
+        x.num_failed_attempts
+        (Date.to_string x.last_failed_attempt)
+        (string_of_client x.client)
+    in
+    match top_3_worst_clients with
+    | [] ->
+        Printf.sprintf
+          "We found a total of %i failed attempts to login since last checking \
+           - these were all of unknown origin!"
+          unknown_client_failed_attempts
+    | _ :: _ ->
+        let worst =
+          List.map string_of_client_failed_attempts top_3_worst_clients
+          |> String.concat "\n"
+        in
+        if unknown_client_failed_attempts > 0 then
+          Printf.sprintf
+            "We found a total of %i failed attempts to login since last \
+             checking, including %i of unknown origin! The worst known \
+             offenders were:\n\
+             %s"
+            total_num_failed_attempts unknown_client_failed_attempts worst
+        else
+          Printf.sprintf
+            "We found a total of %i failed attempts to login since last \
+             checking! The worst offenders were:\n\
+             %s"
+            total_num_failed_attempts worst
```

Comment:
For certificate alerts we used xml to structure messages, if I remember correctly to parse dates independently:

```
 Printf.sprintf "<body><message>%s</message><date>%s</date></body>" msg (Date.to_string expiry)
```

@kc284 What format to display these alerts would be better suited for xencenter? Would xml help you to display the dates in here, while the rest of the message is more freeform?

Reply:
Yes, presumably these numbers need to be shown by the UI, hence isolating them in a separate field would be much easier to parse, otherwise XenCenter has to use a text based regex which will break if in future the wording on the message changes. Json would also work, if you want smaller data. E.g.:
```
<body>
    <offence>
        <message>Total failed attempts to login since last checking</message>
        <number>%i</number>
    </offence>
    <offence>
        <message>Attempts of unknown origin:</message>
        <number>%i</number>
    </offence>
    <offence>
        <message>Worst known offenders</message>
        <number>%s</number>
    </offence>
</body>
```

Reply:
Or maybe use different tag names instead of "offence" so you're not not order bound, e.g. "total", "unknown", "offenders".